collector/src/rustc-fake.rs

```diff
@@ -77,62 +78,62 @@ fn main() {
-                            }
-                            let file = entry.file_name().to_str().unwrap().to_owned();
-                            let new_prefix = Some(file[..file.find('.').unwrap()].to_owned());
-                            assert!(
-                                prefix.is_none() || prefix == new_prefix,
-                                "prefix={:?}, new_prefix={:?}",
-                                prefix,
-                                new_prefix
-                            );
-                            prefix = new_prefix;
-                        }
-                    }
-                    if let Some(profile_data) = full_path {
-                        // measureme 0.8 has a single file
-                        println!("!self-profile-file:{}", profile_data.to_str().unwrap());
-                        let filename = profile_data.file_name().unwrap().to_str().unwrap();
-                        let json = match run_summarize("summarize", &prof_out_dir, filename) {
-                            Ok(s) => s,
-                            Err(e1) => {
-                                match run_summarize("summarize-9.0", &prof_out_dir, filename) {
-                                    Ok(s) => s,
-                                    Err(e2) => {
-                                        panic!("failed to run summarize and summarize-9.0. Errors:\nsummarize: {:?}\nsummarize-9.0: {:?}", e1, e2);
-                                    }
-                                }
-                            }
-                        };
-                        println!("!self-profile-output:{}", json);
-                    } else {
-                        let prefix = prefix.expect(&format!("found prefix {:?}", prof_out_dir));
-                        let json = run_summarize("summarize", &prof_out_dir, &prefix)
-                            .or_else(|_| run_summarize("summarize-0.7", &prof_out_dir, &prefix))
-                            .expect("able to run summarize or summarize-0.7");
-                        println!("!self-profile-dir:{}", prof_out_dir.to_str().unwrap());
-                        println!("!self-profile-prefix:{}", prefix);
-                        println!("!self-profile-output:{}", json);
-                    }
+                    process_self_profile_output(prof_out_dir, &args[..]);
+                }
+            }
+
+            "xperf-stat" | "xperf-stat-self-profile" => {
+                // Read the path to tracelog.exe from either an environment variable, falling back to assuming it's on the PATH.
+                let tracelog = std::env::var("TRACELOG").unwrap_or("tracelog.exe".to_string());
+                let mut cmd = Command::new(tracelog);
+                assert!(cmd.output().is_ok(), "tracelog.exe could not be started");
+
+                cmd.args(&["-start", "counters", "-f", "counters.etl", "-eflag", "CSWITCH+PROC_THREAD+LOADER", "-PMC", "InstructionRetired,TotalCycles:CSWITCH"]);
```

Comment:
Would you mind adding documentation somewhere on how this all works? I.e. what the workflow is, which programs are called in which order, what the flags mean? A link to a good description would be fine too.

Reply:
From playing around with this it looks like the "counters" parameter a globally visible "LoggerName" used for coordinating between the `tracelog` and the `xperf` invocations, right? I think it might be a good idea to generate a more meaningful name here that includes the fact that it belongs to "rustc-perf" and maybe the name of benchmark being run. Something like `rustc-perf-{benchmark}-logger`.

I don't know if it would also make sense to add some randomly generated ID to it too. At least I had some problems where I had to run `xperf -stop counters` manually because otherwise `tracelog` would complain that `counters` already exists. I think my system got into that state because there was some crash after `tracelog` allocated the logger with name "counters" and the `xperf` command stopping the logger was never executed. Making that more robust one way or other seems like a good idea.

Reply:
Added some comments here explaining what the workflow for collecting the counters is. 

I also added code to issue a `xperf -stop` before we do our current run. I think that's better than generating a different name each time as I've read there is a limit to how many concurrent collections you can have running and, if we know the collection name, we can just try stopping it.